### PR TITLE
Implemented node drain.

### PR DESCRIPTION
**Pull request by shreyas-badiger**
Unlike the current version of the controller, in V2 we will be calling drain using the kubectl-pkg-provided-drain rather than calling the drain using `kubectl` binary.

### CODE REVIEW

File: controllers/providers/kubernetes/nodes.go [hunk 1]
```diff
@@ -32,3 +38,45 @@ func (k *KubernetesClientSet) ListClusterNodes() (*corev1.NodeList, error) {
...
+		GracePeriodSeconds:  -1,
+		IgnoreAllDaemonSets: true,
+		Out:                 os.Stdout,
+		ErrOut:              os.Stdout,
+		DeleteEmptyDirData:  true,
+		Timeout:             time.Duration(120) * time.Second,
+	}
+
+	if err := drain.RunCordonOrUncordon(helper, node, true); err != nil {
+		if apierrors.IsNotFound(err) {
```
Comment (eytan-avisror) [inline controllers/providers/kubernetes/nodes.go:64]: Did you verify this returns an IsNotFound error?

File: controllers/providers/kubernetes/nodes.go [hunk 2]
```diff
@@ -32,3 +38,45 @@ func (k *KubernetesClientSet) ListClusterNodes() (*corev1.NodeList, error) {
...
+	}
+
+	if err := drain.RunNodeDrain(helper, node.Name); err != nil {
+		if apierrors.IsNotFound(err) {
+			return err
+		}
+		return fmt.Errorf("error draining node: %v", err)
+	}
+
+	if PostDrainDelaySeconds > 0 {
```
Comment (eytan-avisror) [inline controllers/providers/kubernetes/nodes.go:77]: This should be done via requeue and comparison of timestamp - no more sleeping :)

File: controllers/providers/kubernetes/nodes.go [hunk 3]
```diff
@@ -32,3 +38,45 @@ func (k *KubernetesClientSet) ListClusterNodes() (*corev1.NodeList, error) {
...
+
+	helper := &drain.Helper{
+		Client:              client,
+		Force:               true,
+		GracePeriodSeconds:  -1,
+		IgnoreAllDaemonSets: true,
+		Out:                 os.Stdout,
+		ErrOut:              os.Stdout,
+		DeleteEmptyDirData:  true,
+		Timeout:             time.Duration(120) * time.Second,
```
Comment (eytan-avisror) [inline controllers/providers/kubernetes/nodes.go:60]: What is the timeout we set in v1 controller? can we keep the same?

File: controllers/providers/kubernetes/nodes.go (hunk 2 above)
Comment (eytan-avisror) [reply controllers/providers/kubernetes/nodes.go:77]: We can do this in a separate story where we handle parallelizing this

Comment (shreyas-badiger) [reply controllers/providers/kubernetes/nodes.go:77]: Ah! Yeah. I think we are already doing that as I am setting the lastNodeDrainTime. I shall remove this sleep.

File: controllers/providers/kubernetes/nodes.go (hunk 3 above)
Comment (shreyas-badiger) [reply controllers/providers/kubernetes/nodes.go:60]: Yes. I have to pull that from Spec. I just did

File: controllers/providers/kubernetes/nodes.go (hunk 1 above)
Comment (shreyas-badiger) [reply controllers/providers/kubernetes/nodes.go:64]: So, under apierrors, there are several different type of errors. IsNotFound() returns error when a resource type is not found.
"IsNotFound returns true if the specified error was created by NewNotFound. NewNotFound returns a new error which indicates that the resource of the kind and the name was not found."

However, I wasn't able to catch the error for invalid nodename. I checked the code for RunNodeDrain(),  they do return an error when they can't select pods for eviction, but I didn't see any check for empty list of pods. I mean, when the pod selector returns an empty list, they just move on to the next steps.

**Review by tekenstam — APPROVED** [review]

**Pull request merged**